### PR TITLE
Endpoint scopes

### LINKED ISSUE #277: kpng brain needs CONSISTENT hashing --> cannot rely on golang map iteration due to randomness.... IPVS broken bc of this.
........... (golang intentionally randomly orders map keys) .........

@VivekThrivikraman-est @hanamantagoudvk @nehaLohia27 have been deep diving into IPVS and are finding that ever since the refactoring, ipvs is logically inconsistent/things are breaking. 

IPTables isnt suffering from this failure though b/c it uses the serviceChangeTracker and doesnt need the kpng hashes to be in place.

Lets add more logs and logic to the kpng brain and see exactly whats happening/fix it.

### PR DESCRIPTION
**Pull request by mcluseau**
Introduce the notion of endpoint scopes to allow generic filtering with a simple node-local model.

Includes 2 first scopes: Internal and External.

Depends on #286 
Should fix #277

### CODE REVIEW

File: backends/iptables/iptables.go [hunk 1]
```diff
@@ -256,7 +257,7 @@ func (t *iptables) sync() {
 }
 
 func (t *iptables) createServiceSpecificChains(svcInfo *serviceInfo, activeNATChains map[util.Chain]bool,
-	existingNATChains map[util.Chain][]byte, allEndpoints *endpointsInfoByName) ([]*string, *[]util.Chain, *[]util.Chain, map[string]map[string]int32) {
+	existingNATChains map[util.Chain][]byte, allEndpoints *endpointsInfoByName) ([]*string, *[]util.Chain, *[]util.Chain, map[string]int32) {
```
Comment (jayunit100) [inline backends/iptables/iptables.go:260]: oh so were replacing 

```
str0 -> 
   str1 -> int1
   str2 -> int2
```

w/ 
```
str0 -> int0
str1 -> int1 
```

why ? sorry am reloading my kpng brain been focused on downstream for a few months

**Comment by jayunit100** [comment]
fixes https://github.com/kubernetes-sigs/kpng/issues/277

File: api/localnetv1/endpoint-scopes.go [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package localnetv1
+
+// Any returns true iff there's any scope at true
+func (s *EndpointScopes) Any() bool {
+	return s.Internal ||
```
Comment (jayunit100) [inline api/localnetv1/endpoint-scopes.go:5]: (for clusterTrafficPolicy=internal/external)

**Comment by mcluseau** [comment]
@jayunit100 this need an lgtm, and is probably on the path to solve #312

**Pull request merged**